### ISSUE TITLE
New Feature: Add -u|--user to "docker cp" command

### ISSUE THREAD

**Issue by ArSa-Apps** [docker/docker#25977]
Is it possible to add this option to "docker cp"? When I try to copy files into container, it end up being copied as root which will break the start-up chain (permission denied!) of container and make it really difficult to change the permission.
Cheers & thanks,
A.

**Comment by Frutal1ty** [comment]
I agree with you that would be nice, but it is really that hard to change the owner UID:GID of those files?
I mean, you can do it with:
`docker exec -it container chown user:group files`

Moreover, I don't know how difficult this would be to implement because as far as I understand that would require the docker-engine to have acces to container's user list, and also to the permissions of the destiny folder, to check if the required user can write.
The current behaviour makes all much easier as all linux OS have a root user and hes able to write anywhere.

As I said would be a nice feature but I don't know if it's even profitable.

**Comment by ArSa-Apps** [comment]
Well, here is the issue, if the container is already running, obviously you can change the ownership etc. But in my case, since the permissions are not set right, the container can not even start, so the "`docker exec`" would have no effects what so ever...

**Comment by cpuguy83** [comment]
SGTM

**Comment by Kronos1776** [comment]
The only way this could be done automatically is to have a reference file within the volume that could be used as a permissions template.  Then docker cp could mimic the settings from that file to set the permissions on the files copied in.

This is a very easy thing to do and would not tax the docker dev staff too much to do that as this is all host work and would provide an easy way for admins to get files into a volume without undue fuss.

**Comment by turnerguo** [comment]
The same issue on me.
I think it is possible to just simply override the `DEST_PATH` permission with its `CONTAINER` permission while do `docker cp`, instead of manually modify the permissions.

**Comment by phra** [comment]
i'm stuck too for the same reason. i have a data volume that if it has the wrong permission it blocks another container and because the data volume is not running i can't change them. 😕

**Comment by razic** [comment]
thinking about picking this one up after #8755 -- @cpuguy83 is this still something we'd like?

**Comment by cpuguy83** [comment]
Makes sense to me.

**Comment by thaJeztah** [comment]
Wasn't this worked on in https://github.com/docker/docker/pull/28923 ?

**Comment by Rots** [comment]
I have another use case, which needs this feature...
The dockerd is deployed with authz plugin that lets me run containers (and execute commands) with `--user 1000` but not with `--user root`. I can also use `docker cp` for copying files into the container, but because I cannot change the file permissions, I have issues where the processes expect different permissions.
Yes, there are workarounds:
1) copy files to a `/known/path` into the container, then inside the container do `cp /known/path target/path`. Same problem as in comment https://github.com/moby/moby/issues/25977#issuecomment-242129050
2) use bind mounts or volumes. The problem with this approach is that currently I don't need this data to persisted anywhere outside the container, in fact, it is already coming from another system and I can recreate the files through other means if necessary.
3) I guess as the permission flags are kept, I could try `chmod -R a+wx` before copying the files into the container (not tested yet)

This feature would enable us to use the containers this way.

**Comment by themanifold** [comment]
This would be handy for me too, but a decent workaround would be something like this:
```
docker exec -u root some_container cat /some/file > /some/where/on/your/host
```

**Comment by timothyclarke** [comment]
Firstly why is there a 4 year old issue for important functionality still sitting unaddressed 
in a `Dockerfile` you can pass `--chown <uid>:<gid>` to the `COPY` command so all statements around "It's difficult to figure out the correct ID when copying" are moot, just have a restriction of ID's only.  
With the exception of the response from @themanifold if the container starts as a non-root user (the dockerfile has `USER` in it) then doing a `chown` is not an option. I'm wanting to copy the files from my UID (on my pc) to the "default" UID in the docker container.
